### ISSUE TITLE
Background Music not working properly

### ISSUE THREAD

**Issue by AltyMan** [kyleneideck/BackgroundMusic#266]
So my Background Music does not let me mute specific applications for some reason, but everything else works, including automatic Spotify pausing and turning down the master volume.

**Comment by kode54** [comment]
Maybe it needs the Apple Events entitlement to communicate with other processes that are sending audio.

**Comment by AltyMan** [comment]
> Maybe it needs the Apple Events entitlement to communicate with other processes that are sending audio.

And how do I do that?

**Comment by kode54** [comment]
It's something the developer would likely have to test, if it is indeed something that would fix it.

There's some sort of entitlement or security issue that became a problem as of Catalina, preventing the BackgroundMusic app from communicating with the driver portion that's running within the apps themselves.

Is volume control even working?

**Comment by arimus** [comment]
Same problem here on Catalina. For me, I can control master sound and see all the apps, however changing volume per application doesn't work. I haven't yet tried to do a custom build yet.

**Comment by kode54** [comment]
As I suggested, it's probably the new security entitlement, because by default Apple Events are only allowed to be sent or received within your own apps, or apps with similar base bundle identifiers. I can try something next time I'm in macOS, and see if it works.

**Comment by arimus** [comment]
@kode54 It looks like the project workspace already has Audio Input and Apple Events entitlements added to the hardened runtime settings. I built with and without my personal signing certs, just to see if that would make a different. No go. Also appears that there are no errors when attempting to control the mixer levels for an app:

`2020-03-24 20:22:19.697632-0600 Background Music[77883:4958750] BGMAppVolumes::appVolumeChanged: App volume for com.google.Chrome.canary (74655) changed to 46
2020-03-24 20:22:19.729704-0600 Background Music[77883:4958750] BGMAppVolumes::appVolumeChanged: App volume for com.google.Chrome.canary (74655) changed to 47
2020-03-24 20:22:19.779711-0600 Background Music[77883:4958750] BGMAppVolumes::appVolumeChanged: App volume for com.google.Chrome.canary (74655) changed to 47
2020-03-24 20:22:20.017868-0600 Background Music[77883:4958750] BGMAppVolumes::appVolumeChanged: App volume for com.google.Chrome.canary (74655) changed to 50`

<img width="512" alt="Screen Shot 2020-03-24 at 8 22 36 PM" src="https://user-images.githubusercontent.com/228394/77495273-adef6900-6e0d-11ea-84e4-d78b3cd13207.png">

Willing to try any suggestions to resolve, if you have any. I don't have time to delve too deeply into the codebase atm and OSX audio is a bit out of my development forte, but I can serve as a pair of hands to monkey around with a few settings / logging / etc. if you point me in the right place ;)

Cheers

**Comment by greatwaldini** [comment]
I've been experiencing this same issue on Catalina, tried installing with pkg as well as homebrew.  Tried all the troubleshooting through mac audio settings, all the basic stuff.  Something seems to be not working right.  Im still able to control volume via system volume.  The BGM app appears to be working in toolbar.  Dropping it down/open reveals the various apps and the volume controller, however, adjusting the volume of any of the apps from BGM has no effect.

**Comment by yurkomik** [comment]
Same situation. Catalina had some changes. OBS also has audio issues. It can't capture background audio any more.

**Comment by kyleneideck** [comment]
@arimus App volumes don't work for some apps and Chrome is one of them. Some apps send a different bundle ID to Core Audio, so BGMDriver (the virtual audio driver) can't match the app to its audio. The plan is to eventually search through the process hierarchy (see #42), but we've been working around the problem with a hardcoded list of apps' bundle IDs and the bundle IDs those apps send to Core Audio.

I've been working on a fix for Chrome (also temporary), but it's not ready yet. Happy to push it to a branch if you want to take a look.

>So my Background Music does not let me mute specific applications for some reason

@AltyMan Which applications? Can you just not mute them or can you not change their volume at all?

>Maybe it needs the Apple Events entitlement to communicate with other processes that are sending audio.

I don't think that would be the problem, unfortunately. Background Music doesn't need any Apple Events permissions from the user to control app volumes. It needs them for auto-pausing in most cases, but it applies the app volumes to the audio the apps send to the Background Music device (in BGMDriver).

**Comment by arimus** [comment]
Happy to test a branch when you get something you need eyes on. Just let me
know.

**Comment by AltyMan** [comment]
> @arimus App volumes don't work for some apps and Chrome is one of them. Some apps send a different bundle ID to Core Audio, so BGMDriver (the virtual audio driver) can't match the app to its audio. The plan is to eventually search through the process hierarchy (see #42), but we've been working around the problem with a hardcoded list of apps' bundle IDs and the bundle IDs those apps send to Core Audio.
> 
> I've been working on a fix for Chrome (also temporary), but it's not ready yet. Happy to push it to a branch if you want to take a look.
> 
> > So my Background Music does not let me mute specific applications for some reason
> 
> @AltyMan Which applications? Can you just not mute them or can you not change their volume at all?
> 
> > Maybe it needs the Apple Events entitlement to communicate with other processes that are sending audio.
> 
> I don't think that would be the problem, unfortunately. Background Music doesn't need any Apple Events permissions from the user to control app volumes. It needs them for auto-pausing in most cases, but it applies the app volumes to the audio the apps send to the Background Music device (in BGMDriver).

Yeah the apps that don’t work for me are Chrome and Firefox, Chrome I understand now but what should I do about Firefox?
